### PR TITLE
Fix syntax for declaring multiple generic type constraints

### LINKED ISSUE #125: Incorrect syntax when declaring multiple constraints for a generic type parameter
`TypeVariableName("T", Comparator::class, Serializable::class)` declared for a function will produce the following signature:

```kotlin
fun <T : Comparator & Serializable> getComparator(): T
```

which doesn't compile. Correct Kotlin syntax is:

```kotlin
fun <T> getComparator(): T where T : Comparator, T : Serializable
```

### PR DESCRIPTION
**Pull request by Egorand**
Fixes #125. Will this need more unit tests? Can you think of any specific scenarios?

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/CodeWriter.kt [hunk 1]
```diff
@@ -158,22 +158,44 @@ internal class CodeWriter @JvmOverloads constructor(
...
+      if (typeVariable.bounds.size > 1) {
+        withMultipleBounds += typeVariable
+      } else if (typeVariable.bounds.size == 1) {
+        emitCode(" : %T", typeVariable.bounds[0])
       }
       firstTypeVariable = false
     }
     emit(">")
+    emitSignature?.invoke()
+    if (withMultipleBounds.isNotEmpty()) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/CodeWriter.kt]: would this be simpler as two calls? One call `emitTypeVariables()` and another, `emitWhereBlock()` ? Custom control flow is a very powerful construct, and I prefer to use simpler constructs wherever we can

File: src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 2]
```diff
@@ -77,10 +77,39 @@ class FunSpec private constructor(builder: Builder) {
     }
 
     if (typeVariables.isNotEmpty()) {
-      codeWriter.emitTypeVariables(typeVariables)
-      codeWriter.emit(" ")
+      codeWriter.emitTypeVariables(typeVariables) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: ... this 8 line block would become a different 6 line block, with no repeated call to `emitSignature`:

```
if (typeVariables.isNotEmpty()) {
  codeWriter.emitTypeVariables(typeVariables);
  codeWriter.emit(" ");
}
emitSignature(codeWriter, enclosingName);
codeWriter.emitWhereBlock(typeVariables);
```

I think it’s more boring an easier to follow

File: src/test/java/com/squareup/kotlinpoet/TypeVariableNameTest.kt [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.kotlinpoet
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import java.io.Serializable
+
+class TypeVariableNameTest {
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/TypeVariableNameTest.kt]: Nice tests

File: src/test/java/com/squareup/kotlinpoet/TypeVariableNameTest.kt [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        .addTypeVariable(TypeVariableName("T", Serializable::class, Runnable::class))
+        .returns(TypeVariableName("T").asNullable())
+        .addStatement("return null")
+        .build()
+    assertThat(funSpec.toString()).isEqualTo("""
+      |fun <T> foo(): T? where T : java.io.Serializable, T : java.lang.Runnable = null
+      |""".trimMargin())
+  }
+
+  @Test fun twoTypeVariablesTwoBoundsEach() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/TypeVariableNameTest.kt]: supercool

**Comment by Egorand** [comment]
Thanks for the feedback @swankjesse!
I wasn't sure about the API, passing `emitSignature()` as a block looked a bit fishy. The solution you suggest is good, but the downside is that currently every spec that needs to print type variables has to call both `emitTypeVariables()` and `emitWhereBlock()` - you can see that I had to modify it in both `FunSpec` and `TypeSpec`. I updated the KDoc to make this dependency clear, but I think it still isn't ideal. Any ideas?

File: src/main/java/com/squareup/kotlinpoet/CodeWriter.kt [hunk 5]
```diff
@@ -166,16 +168,35 @@ internal class CodeWriter @JvmOverloads constructor(
...
     emit(">")
   }
 
+  /**
+   * Emit a `where` block containing type bounds for each type variable that has at least two
+   * bounds.
+   */
+  @Throws(IOException::class)
+  fun emitWhereBlock(typeVariables: List<TypeVariableName>) {
+    if (typeVariables.isEmpty()) return
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/CodeWriter.kt]: this clause isn’t necessary, right?

**Comment by swankjesse** [comment]
Because this is just implementation details – not public API – we can start with this approach which is simpler but requires more careful use. If we want to change our minds later, we can.

For public APIs its much more interesting!

File: src/main/java/com/squareup/kotlinpoet/CodeWriter.kt (hunk 5 above)
Comment (Egorand) [inline src/main/java/com/squareup/kotlinpoet/CodeWriter.kt]: Yep, not really

**Pull request merged**